### PR TITLE
Adds a weighted mode call

### PR DESCRIPTION
**Pull request by rofinn**
Seems like there's a bit of repeated code between other `mode` methods and the `counts` code, but I've attempted to keep these changes isolated to what I need for now.

This PR would let me remove this https://github.com/invenia/Impute.jl/blob/master/src/utils.jl#L31

### CODE REVIEW

**Comment by mschauer** [comment]
LGTM

File: src/scalarstats.jl [hunk 1]
```diff
@@ -158,6 +158,49 @@ function modes(a)
     return [x for (x, c) in cnts if c == mc]
 end
 
+# Weighted mode of arbitrary vectors of values
+function mode(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    isempty(a) && throw(ArgumentError("mode is not defined for empty collections"))
+    length(a) == length(wv) || throw(ArgumentError(
+        "data and weight vectors must be the same size, got $(length(a)) and $(length(wv))"
+    ))
```
Comment (nalimilan) [inline src/scalarstats.jl:166]:
```suggestion
    length(a) == length(wv) ||
        throw(ArgumentError("data and weight vectors must be the same size, got $(length(a)) and $(length(wv))"))
```

File: src/scalarstats.jl [hunk 2]
```diff
@@ -158,6 +158,49 @@ function modes(a)
...
+    end
+
+    return mv
+end
+
+function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    isempty(a) && throw(ArgumentError("mode is not defined for empty collections"))
+    length(a) == length(wv) || throw(ArgumentError(
+        "data and weight vectors must be the same size, got $(length(a)) and $(length(wv))"
+    ))
```
Comment (nalimilan) [inline src/scalarstats.jl:188]:
```suggestion
    length(a) == length(wv) ||
        throw(ArgumentError("data and weight vectors must be the same size, got $(length(a)) and $(length(wv))"))
```

File: test/scalarstats.jl [hunk 3]
```diff
@@ -44,10 +44,23 @@ using Statistics
 @test modes(skipmissing([1, missing, missing, 3, 2, 2, missing])) == [2]
 @test sort(modes(skipmissing([1, missing, 3, 3, 2, 2, missing]))) == [2, 3]
 
+d1 = [1, 2, 3, 3, 4, 5, 5, 3]
+d2 = ['a', 'b', 'c', 'c', 'd', 'e', 'e', 'c']
+wv = weights([0.1:0.1:0.7..., 0.1])
```
Comment (nalimilan) [inline test/scalarstats.jl:49]:
```suggestion
wv = weights([0.1:0.1:0.7; 0.1])
```

File: test/scalarstats.jl [hunk 4]
```diff
@@ -44,10 +44,23 @@ using Statistics
...
+@test mode(d1) == 3
+@test mode(d2) == 'c'
+@test mode(d1, wv) == 5
+@test mode(d2, wv) == 'e'
+@test sort(modes(d1[1:end-1], weights(ones(7)))) == [3, 5]
```
Comment (nalimilan) [inline test/scalarstats.jl:54]: Better test with a non-constant vector?

**Review by nalimilan — COMMENTED** [review]
Thanks, looks mostly good.

(For reference, we don't use `countmap` because it requires iterating over the dict to find what's the largest value. This could probably be fixed by allowing the `countmap` implementation do keep track of the highest value, but that can be left for later.)

File: test/scalarstats.jl (hunk 3 above)
Comment (rofinn) [reply test/scalarstats.jl:49]: Nice, I didn't realize that worked 👍

File: test/scalarstats.jl (hunk 4 above)
Comment (rofinn) [reply test/scalarstats.jl:54]: How about this?

```suggestion
@test sort(modes(d1[1:end-1], weights(ones(7)))) == [3, 5]
@test sort(modes(d1, weights([.9, .1, .1, .1, .9, .1, .1, .1]))) == [1, 4]
```

Comment (nalimilan) [reply test/scalarstats.jl:54]: Sure.

**Comment by rofinn** [comment]
Hmm, I guess this PR need the #621 to be merged?

**Comment by nalimilan** [comment]
Unfortunately yes.

**Review by nalimilan — APPROVED** [review]

**Review by nalimilan — CHANGES_REQUESTED** [review]
Sorry I forgot that the docstrings need to mention the new methods.

**Comment by rofinn** [comment]
macOS 1.0 failing test seems unrelated?

**Comment by mschauer** [comment]
This is a bit untypical for a random failure it is also weight related code (fweight) in https://github.com/JuliaStats/StatsBase.jl/blob/217da39c5a5a6193af75bc03bc3b9f4af0cf212d/test/moments.jl#L168

File: src/scalarstats.jl [hunk 5]
```diff
@@ -47,10 +47,11 @@ end
 # compute mode, given the range of integer values
 """
     mode(a, [r])
+    mode(a, wv)
```
Comment (nalimilan) [inline src/scalarstats.jl:50]:
```suggestion
    mode(a::AbstractArray, wv::AbstractArray{<: Real})
```

File: src/scalarstats.jl [hunk 6]
```diff
@@ -75,9 +76,10 @@ end
 
 """
     modes(a, [r])::Vector
+    modes(a, wv)::Vector
```
Comment (nalimilan) [inline src/scalarstats.jl:79]:
```suggestion
    mode(a::AbstractArray, wv::AbstractArray{<: Real})::Vector
```

File: src/scalarstats.jl (hunk 1 above)
Comment (nalimilan) [reply src/scalarstats.jl:168]: This wasn't actually applied.

**Comment by nalimilan** [comment]
I think it's just bad luck. https://github.com/JuliaStats/StatsBase.jl/pull/624 is an attempt at fixing this.

File: src/scalarstats.jl (hunk 1 above)
Comment (rofinn) [reply src/scalarstats.jl:168]: Oh, no it was, but then I forgot that I applied it on github, so my rebase with master lossed it.

File: test/scalarstats.jl (hunk 4 above)
Comment (rofinn) [reply test/scalarstats.jl:54]:
```suggestion
@test sort(modes(d1[1:end-1], weights(ones(7)))) == [3, 5]
@test sort(modes(d1, weights([.9, .1, .1, .1, .9, .1, .1, .1]))) == [1, 4]
```

File: src/scalarstats.jl [hunk 7]
```diff
@@ -75,9 +76,10 @@ end
 
 """
     modes(a, [r])::Vector
+    mode(a::AbstractArray, wv::AbstractArray{<: Real})::Vector
```
Comment (nalimilan) [inline src/scalarstats.jl:79]:
```suggestion
    mode(a::AbstractArray, wv::AbstractWeights)::Vector
```

File: src/scalarstats.jl [hunk 8]
```diff
@@ -47,10 +47,11 @@ end
 # compute mode, given the range of integer values
 """
     mode(a, [r])
+    mode(a::AbstractArray, wv::AbstractArray{<: Real})
```
Comment (nalimilan) [inline src/scalarstats.jl:50]: Sorry I got this wrong! :-D
```suggestion
    mode(a::AbstractArray, wv::AbstractWeights)
```

**Review by nalimilan — APPROVED** [review]

**Pull request merged**